[API] Indicate maximum attachment file size

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/20858
## Description

The main purpose of this PR is to allow the client to obtain the maximum file size for uploaded attachments.
However I also took the liberty to extend (and document) the information available via the root endpoint (`/api/v3`) a little bit.

NobodysNightmare committed:
add link for adding attachments to work package

NobodysNightmare committed:
add configuration endpoint to APIv3

NobodysNightmare committed:
update root representer

- add it to the documentation
- only provide links to resources that can be directly used (i.e. without templating)
- provide useful general instance information

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-api-attachment-size.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-api-attachment-size.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -461,6 +461,57 @@ Cursor based pagination is therefore less suited for use cases where you want to
+                        "href" : "/api/v3/types"
+                    }
+                },
+                "instanceName" : "Chili con Project",
```
@NobodysNightmare you are aware of the resemblance to the name of one of OP's ancestors?

NobodysNightmare replied:
lol not really, I just used the output of my test instance which features projects like "chili con carne"...

You think I should change that? :D

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -461,6 +461,57 @@ Cursor based pagination is therefore less suited for use cases where you want to
++ Response 200 (application/hal+json)
+
+    [Root][]
+
```
Not really scoped to this PR but when skimming over the documentation I couldn't find a reference to a 401 response (except for https://github.com/NobodysNightmare/openproject/blob/feature/api_attachment_size/doc/apiv3-documentation.apib#L217). I think we should add a general section to the documentation (e.g. https://github.com/NobodysNightmare/openproject/blob/feature/api_attachment_size/doc/apiv3-documentation.apib#L41) that says "The API will always return a 401 if you are not authenticated"

NobodysNightmare replied:
"The API will always return a 401 if you are not authenticated, _and authentication is required_"

That is probably a good idea, but there is more to find on common errors that are not always documented.
I extended the docs on **HTTP 404** lately, but I couldn't promise that I covered all resources.

ulferts commented on lib/api/v3/configuration/configuration_representer.rb:
```diff
@@ -0,0 +1,54 @@
+  module V3
+    module Configuration
+      class ConfigurationRepresenter < ::API::Decorators::Single
+
```
How about getting rid of this @NobodysNightmare?

ulferts commented on spec/lib/api/v3/root_representer_spec.rb:
```diff
@@ -29,35 +29,45 @@
+    end
+
+    it 'indicates the OpenProject version number' do
+      is_expected.to be_json_eql(Redmine::VERSION.to_semver.to_json).at_path('coreVersion')
```
Ah ... this PR is full of nostalgia. 

You can use `OpenProject::VERSION.to_semver.to_json` just the same.

But I think it would be even better, to mock this.

ulferts commented:
Only minor adjustments necessary, @NobodysNightmare. Fix those and this is good to merge.

NobodysNightmare committed:
replace references to Redmine::VERSION

- we now live in OpenProject for quite some time
- kept legacy specs because they are... legacy

NobodysNightmare committed:
document general HTTP 401 behaviour of APIv3
